src/directives.ts

```diff
@@ -0,0 +1,138 @@
+import { VNode, VueConstructor } from 'vue';
+import { DirectiveBinding, DirectiveOptions } from 'vue/types/options';
+
+import { getClosest } from './utils';
+
+
+interface DestroyHTMLElement extends HTMLElement {
+    $destroy: () => void;
+}
+
+interface DestroyHTMLInputElement extends HTMLInputElement {
+  $destroy: () => void;
+}
+
+/**
+ * Triggers action when scroll gets close to the bottom.
+ *
+ * Example usage:
+ * <div v-infinite-scroll="action()">Scroll me</div>
+ */
+const infiniteScroll: DirectiveOptions = {
+  bind(el: DestroyHTMLElement, binding: DirectiveBinding): void {
+    const tolerance = parseInt(binding.arg || '100', 10);
+    const action = binding.value;
+
+    el.addEventListener('scroll', (): void => {
+      const height = (el.firstChild as DestroyHTMLElement).clientHeight;
+      if (el.scrollTop >= height - el.clientHeight - tolerance) {
+        action();
+      }
+    });
+  },
+};
+
+/**
+ * Scroll input into view when focused.
+ *
+ * Example usage:
+ * <input type="text" value="foo" v-scroll-on-focus />
+ */
+const scrollOnFocus: DirectiveOptions = {
+  bind(el: DestroyHTMLInputElement): void {
+    const target = el.querySelector('input,textarea');
+    if (!target) {
+      return;
+    }
+
+    const scrollTo = (target: HTMLElement, scrollable: HTMLElement, scrollableContent: HTMLElement, marginTop: number = 0): void => {
+      while (target.offsetParent && target.offsetParent !== scrollable) {
+        marginTop -= target.offsetTop;
+        target = target.offsetParent as HTMLElement;
+      }
+      scrollableContent.scrollTop = target.offsetTop - marginTop;
+    };
+
+    const handler = (): void => {
+      const scrollable = getClosest(el, '.scrollable');
+      const scrollableContent = getClosest(el, '.scrollable-content');
+      if (scrollable && scrollableContent) {
+        const height = scrollableContent.offsetHeight;
+        setTimeout((): void => {
+          // if scrollableContent height is reduced in half second
+          // since an input got focus we assume soft keyboard is showing.
+          if (height > scrollableContent.offsetHeight) {
+            scrollTo(target as HTMLElement, scrollable, scrollableContent, 20);
+          }
+        }, 500);
+      }
+    };
+
+    target.addEventListener('focus', handler);
+    el.$destroy = (): void => {
+      target.removeEventListener('focus', handler);
+    };
+  },
+  unbind(el: DestroyHTMLElement): void {
+    el.$destroy();
+  },
+};
+
+export interface TrackClickDataAttributes {
```

Comment:
I think it'd be more concise/clearer to just make the function signature include this type information. So instead of taking an object of this type as an input, trackEvent would just have a signature of `(eventName: string, eventProps: {[key]: string: any})`

Reply:
yes. That is better :)